pkg/controller/master/image/vm_image_controller.go

```diff
@@ -0,0 +1,160 @@
+package image
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	"github.com/longhorn/longhorn-manager/types"
+	v1 "github.com/rancher/wrangler/pkg/generated/controllers/storage/v1"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	ctlharvesterv1 "github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+	lhv1beta1 "github.com/harvester/harvester/pkg/generated/controllers/longhorn.io/v1beta1"
+	"github.com/harvester/harvester/pkg/ref"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	optionBackingImageName = "backingImage"
+	optionMigratable       = "migratable"
+)
+
+// vmImageHandler syncs status on vm image changes, and manage a storageclass & a backingimage per vm image
+type vmImageHandler struct {
+	httpClient     http.Client
+	storageClasses v1.StorageClassClient
+	images         ctlharvesterv1.VirtualMachineImageClient
+	backingImages  lhv1beta1.BackingImageClient
+}
+
+func (h *vmImageHandler) OnChanged(key string, image *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+	if image == nil || image.DeletionTimestamp != nil {
+		return image, nil
+	}
+	if !harvesterv1.ImageInitialized.IsTrue(image) {
+		return h.initialize(image)
+	} else if image.Spec.URL != image.Status.AppliedURL {
+		// URL is changed, recreate the storageclass and backingimage
+		if err := h.backingImages.Delete(util.LonghornSystemNamespace, getBackingImageName(image), &metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+			return image, err
+		}
+		if err := h.storageClasses.Delete(getImageStorageClassName(image.Name), &metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+			return image, err
+		}
+		return h.initialize(image)
+	}
+	return image, nil
+}
+
+func (h *vmImageHandler) OnRemove(_ string, image *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+	if image == nil {
+		return nil, nil
+	}
+	scName := getImageStorageClassName(image.Name)
+	if err := h.storageClasses.Delete(scName, &metav1.DeleteOptions{}); !errors.IsNotFound(err) && err != nil {
+		return image, err
+	}
+	biName := getBackingImageName(image)
+	if err := h.backingImages.Delete(util.LonghornSystemNamespace, biName, &metav1.DeleteOptions{}); !errors.IsNotFound(err) && err != nil {
+		return image, err
+	}
+	return image, nil
+}
+
+func (h *vmImageHandler) initialize(image *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+	if err := h.createBackingImage(image); err != nil && !errors.IsAlreadyExists(err) {
+		return nil, err
+	}
+	if err := h.createStorageClass(image); err != nil && !errors.IsAlreadyExists(err) {
+		return nil, err
+	}
```

Comment:
can we add a proper error message to the failed initialization?

Reply:
The controller is expected to reconcile with these errors so we don't set initialization as failed here.